language/move-vm/runtime/src/interpreter.rs

```diff
@@ -262,15 +262,38 @@ impl Interpreter {
     ///
     /// Native functions do not push a frame at the moment and as such errors from a native
     /// function are incorrectly attributed to the caller.
-    fn make_call_frame(&mut self, func: Arc<Function>, ty_args: Vec<Type>) -> VMResult<Frame> {
+    fn make_call_frame<'a>(
+        &mut self,
+        func: Arc<Function>,
+        ty_args: Vec<Type>,
+        loader: &'a Loader,
+    ) -> VMResult<Frame> {
+        let resolver = func.get_resolver(loader);
         let mut locals = Locals::new(func.local_count());
         let arg_count = func.arg_count();
         for i in 0..arg_count {
+            let v = self.operand_stack.pop().map_err(|e| self.set_location(e))?;
+            let expected_ty = resolver
+                .resolve_signature_token(&func.local_types()[arg_count - i - 1])
+                .and_then(|ty| ty.subst(&ty_args))
+                .map_err(|e| match func.module_id() {
+                    Some(id) => e
+                        .at_code_offset(func.index(), 0)
+                        .finish(Location::Module(id.clone())),
+                    None => {
+                        let err =
+                            PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                                .with_message(
+                                    "Unexpected native function not located in a module".to_owned(),
+                                );
+                        self.set_location(err)
+                    }
+                })?;
+
+            v.check_type(&expected_ty)
```

Comment:
I think it makes more sense to check the type at each instruction. For calls, that will cover both the argument and return types, whereas this only covers the argument types.

Reply:
I feel like this would be a much greater change than this PR: this mean that we want to have a typed operand stack which could be hard to construct with the current design, as the VM right now doesn't carry any type info with instruction which made it quite hard to check. I think maybe we can get started from checking the following three locations:

1. Function Call
2. WriteRef(not in this PR)
3. Write to storage

I think these should offer us enough protection over unexpected type conversion.